# Conteúdo <a id='back'></a>

* [Introdução](#intro)
* [Etapa 1. Visão geral dos dados](#data_review)
    * [Conclusões](#data_review_conclusions)
* [Etapa 2. Pré-processamento de dados](#data_preprocessing)
    * [2.1 Estilo do cabeçalho](#header_style)
    * [2.2 Valores ausentes](#missing_values)
    * [2.3 Duplicados](#duplicates)
    * [2.4 Conclusões](#data_preprocessing_conclusions)
* [Etapa 3. Teste da hipótese](#hypothesis)
    * [3.1 Hipótese 1: atividade dos usuários nas duas cidades](#activity)
* [Conclusões](#end)

## Introdução <a id='intro'></a>
O trabalho de um analista é analisar dados para obter percepções valiosas dos dados e tomar decisões fundamentadas neles. Esse processo consiste em várias etapas, como visão geral dos dados, pré-processamento dos dados e testes de hipóteses.

Sempre que fazemos uma pesquisa, precisamos formular uma hipótese que depois poderemos testar. Às vezes nós aceitamos essas hipóteses; outras vezes, nós as rejeitamos. Para fazer as escolhas certas, um negócio deve ser capaz de entender se está fazendo as suposições certas ou não.

Neste projeto, você vai comparar as preferências musicais dos habitantes de Springfild e Shelbyville. Você vai estudar os dados de um serviço de streaming de música online para testar a hipótese apresentada abaixo e comparar o comportamento dos usuários dessas duas cidades.

### Objetivo:
Teste a hipótese:
1. A atividade dos usuários é diferente dependendo do dia da semana e da cidade.


### Etapas
Os dados sobre o comportamento do usuário são armazenados no arquivo `/datasets/music_project_en.csv`. Não há informações sobre a qualidade dos dados, então será necessário examiná-los antes de testar a hipótese.

Primeiro, você avaliará a qualidade dos dados e verá se seus problemas são significativos. Depois, durante o pré-processamento dos dados, você tentará tratar dos problemas mais críticos.

O seu projeto consistirá em três etapas:
 1. Visão geral dos dados
 2. Pré-processamento de dados
 3. Teste da hipótese








[Voltar ao Índice](#back)

## Etapa 1. Visão geral dos dados <a id='data_review'></a>

In [2]:
# importando pandas
import pandas as pd

Leia o arquivo `music_project_en.csv` da pasta `/datasets/` e salve-o na variável `df`:

In [3]:
# lendo o arquivo e armazenando em df
df = pd.read_csv('..\dataset\music_project_en.csv')

Imprima as primeiras 10 linhas da tabela:

In [5]:
# obtenha as 10 primeiras 10 linhas da tabela df
print(df.head(10))

     userID                        Track            artist   genre  \
0  FFB692EC            Kamigata To Boots  The Mass Missile    rock   
1  55204538  Delayed Because of Accident  Andreas Rönnberg    rock   
2    20EC38            Funiculì funiculà       Mario Lanza     pop   
3  A3DD03C9        Dragons in the Sunset        Fire + Ice    folk   
4  E2DC1FAE                  Soul People        Space Echo   dance   
5  842029A1                       Chains          Obladaet  rusrap   
6  4CB90AA5                         True      Roman Messer   dance   
7  F03E1C1F             Feeling This Way   Polina Griffith   dance   
8  8FA1D3BE                     L’estate       Julia Dalia  ruspop   
9  E772D5C0                    Pessimist               NaN   dance   

        City        time        Day  
0  Shelbyville  20:28:33  Wednesday  
1  Springfield  14:07:09     Friday  
2  Shelbyville  20:58:07  Wednesday  
3  Shelbyville  08:37:09     Monday  
4  Springfield  08:34:34     Monday  
5

Obtenha informações gerais sobre a tabela usando um comando. Você conhece o método para exibir informações gerais que precisamos obter.

In [6]:
# obtendo informações gerais sobre os nossos dados
print(df.info())

<class 'pandas.DataFrame'>
RangeIndex: 65079 entries, 0 to 65078
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype
---  ------    --------------  -----
 0     userID  65079 non-null  str  
 1   Track     63736 non-null  str  
 2   artist    57512 non-null  str  
 3   genre     63881 non-null  str  
 4     City    65079 non-null  str  
 5   time      65079 non-null  str  
 6   Day       65079 non-null  str  
dtypes: str(7)
memory usage: 3.5 MB
None


Aqui estão as nossas observações sobre a tabela. Ela contém sete colunas. Elas armazenam o mesmo tipo de dado: `object`.

De acordo com a documentação:
- `' userID'` — identificação do usuário
- `'Track'` — título da música
- `'artist'` — nome do artista
- `'genre'` — gênero da música
- `'City'` — cidade do usuário
- `'time'` — o tempo exato que a música foi reproduzida
- `'Day'` — dia da semana

Podemos ver três problemas de estilo nos cabeçalhos da tabela:
1. Alguns cabeçalhos são escritos em letras maiúsculas, outros estão em minúsculas.
2. Alguns cabeçalhos contêm espaços.
3. Um dos cabeçalhos possui dois substantivos, que estão concatenados e não separados por underscore de acordo com as principais convenções de Python.




Os dados presentes na tabela são do tipo objeto (string), o que pode se tornar um problema quando esta formatação está presente em algorismos.

Sem uma análise profunda, é inconclusivo a resposta para hipótese

Sim, pela análise das informações do DF, foi possível constatar a presença de valores ausentes.

[Voltar ao Índice](#back)

## Etapa 2. Pré-processamento de dados <a id='data_preprocessing'></a>

O objetivo aqui é preparar os dados para a análise.
O primeiro passo é resolver todos os problemas com o cabeçalho. E então podemos passar para os valores ausentes e duplicados. Vamos começar.

Corrija a formatação nos cabeçalhos da tabela.


### Estilo do cabeçalho <a id='header_style'></a>
Imprima os cabeçalhos da tabela (os nomes das colunas):

In [7]:
# imprima os nomes das colunas
print(df.columns)

Index(['  userID', 'Track', 'artist', 'genre', '  City  ', 'time', 'Day'], dtype='str')


Mude os cabeçalhos da tabela conforme as boas práticas de estilo:
* Todos os caracteres precisam estar com letras minúsculas
* Exclua espaços
* Se o nome tiver várias palavras, use snake_case

Anteriormente, você aprendeu sobre uma maneira automatizada de renomear colunas. Vamos usá-la agora. Use o ciclo for para percorrer os nomes das colunas e transformar todos os caracteres em letras minúsculas. Após fazer isso, imprima os cabeçalhos da tabela novamente:

In [8]:
# Percorrendo os cabeçalhos e convertendo tudo em minúsculos
old_names = df.columns
new_names = list()
for old_name in old_names:
    lowered = old_name.lower()
    new_names.append(lowered)
df.columns = new_names
print(df.columns)

Index(['  userid', 'track', 'artist', 'genre', '  city  ', 'time', 'day'], dtype='str')


Agora, usando a mesma abordagem, exclua os espaços no início e no final de cada nome de coluna e imprima os nomes das colunas novamente:

In [9]:
# Percorrendo os cabeçalhos e removendo os espaços
old_names = df.columns
new_names = list()
for old_name in old_names:
    striped = old_name.strip()
    new_names.append(striped)
df.columns = new_names
print(df.columns)

Index(['userid', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='str')


Precisamos aplicar a regra de sublinhado no lugar de espaço à coluna `userid`. Deveria ser `user_id`. Renomeie essa coluna e imprima os nomes de todas as colunas quando terminar.

In [10]:
# Renomeando a coluna "userid"
df.rename(columns = {'userid' : 'user_id'}, inplace=True)
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='str')


Verifique o resultado. Imprima os cabeçalhos novamente:

In [11]:
# verificando o resultado: a lista de cabeçalhos
print(df.columns)

Index(['user_id', 'track', 'artist', 'genre', 'city', 'time', 'day'], dtype='str')


[Voltar ao Índice](#back)

### Valores Ausentes <a id='missing_values'></a>
 Primeiro, encontre a quantidade de valores ausentes na tabela. Você precisa usar dois métodos em sequência para obter o número de valores ausentes.

In [12]:
# calculando o número de valores ausentes
print(df.isna().sum())

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64


Nem todos os valores ausentes afetam a pesquisa. Por exemplo, os valores ausentes em `track` e `artist` não são críticos. Você pode simplesmente substituí-los por valores padrão, como a string `'unknown'`.

Mas valores ausentes em `'genre'` podem afetar a comparação de preferências musicais de Springfield e Shelbyville. Na vida real, seria útil descobrir as razões pelas quais os dados estão ausentes e tentar corrigi-los. Mas nós não temos essa possibilidade neste projeto. Então, você terá que:
* Preencha esses valores ausentes com um valor padrão
* Avalie em que medida os valores ausentes podem afetar sua análise

Substitua os valores ausentes nas colunas `'track'`, `'artist'` e `'genre'` pela string `'unknown'`. Como mostramos nas lições anteriores, a melhor maneira de fazer isso é criar uma lista para armazenar os nomes das colunas nas quais precisamos fazer a substituição. Em seguida, use essa lista e percorra as colunas nas quais a substituição seja necessária e faça a substituição.

In [13]:
# percorrendo os cabeçalhos e substituindo valores ausentes por 'unknown'
columns_replacing = ['artist', 'genre', 'track']
for col in columns_replacing:
    df[col].fillna('unknown', inplace=True)

C:\Users\Brian\AppData\Local\Temp\ipykernel_6960\622490109.py:4: ChainedAssignmentError: A value is being set on a copy of a DataFrame or Series through chained assignment using an inplace method.
Such inplace method never works to update the original DataFrame or Series, because the intermediate object on which we are setting values always behaves as a copy (due to Copy-on-Write).

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' instead, to perform the operation inplace on the original object, or try to avoid an inplace operation using 'df[col] = df[col].method(value)'.

See the documentation for a more detailed explanation: https://pandas.pydata.org/pandas-docs/stable/user_guide/copy_on_write.html
  df[col].fillna('unknown', inplace=True)


Agora verifique o resultado para ter certeza de que o conjunto de dados não contenha valores ausentes após a substituição. Para fazer isso, conte os valores ausentes novamente.

In [14]:
# contando os valores ausentes
df.isna().sum()

user_id       0
track      1343
artist     7567
genre      1198
city          0
time          0
day           0
dtype: int64

[Voltar ao Índice](#back)

### Duplicados <a id='duplicates'></a>
Encontre o número de duplicados explícitos na tabela. Lembre-se de que você precisa aplicar dois métodos em sequência para obter o número de duplicados explícitos.

In [15]:
# contando duplicados explícitos
print(df.duplicated().sum())

3826


Agora descarte todos os duplicados. Para fazer isso, chame o método que faz exatamente isso.

In [16]:
# removendo duplicados explícitos
df = df.drop_duplicates().reset_index(drop=True)

Agora vamos verificar se descartamos todos os duplicados. Conte duplicados explícitos mais uma vez para ter certeza de que você removeu todos eles:

In [17]:
# verificando duplicados novamente
print(df.duplicated().sum())

0


Agora queremos nos livrar dos duplicados implícitos na coluna `genre`. Por exemplo, o nome de um gênero pode ser escrito de maneiras diferentes. Alguns erros afetarão também o resultado.

Para fazer isso, vamos começar imprimindo uma lista de nomes de gênero únicos, ordenados em ordem alfabética: Para fazer isso:
* Extraia a coluna `genre` do DataFrame
* Chame o método que retornará todos os valores únicos na coluna extraída


In [18]:
# visualizando nomes de gêneros únicos
print(df['genre'].unique())

<StringArray>
[       'rock',         'pop',        'folk',       'dance',      'rusrap',
      'ruspop',       'world',  'electronic',           nan, 'alternative',
 ...
     'indipop',         'axé',        'fado',   'showtunes',       'arena',
       'irish',    'mandopop',       'forró',       'dirty',    'regional']
Length: 269, dtype: str


Olhe a lista e encontre duplicados implícitos do gênero `hiphop`. Esses podem ser nomes escritos incorretamente, ou nomes alternativos para o mesmo gênero.

Você verá os seguintes duplicados implícitos:
* `hip`
* `hop`
* `hip-hop`

Para se livrar deles, crie uma função `replace_wrong_genres()` com dois parâmetros:
* `wrong_genres=` — essa é uma lista que contém todos os valores que você precisa substituir
* `correct_genre=` — essa é uma string que você vai usar para a substituição

Como resultado, a função deve corrigir os nomes na coluna `'genre'` da tabela `df`, isto é, substituindo cada valor da lista `wrong_genres` por valores de `correct_genre`.

Dentro do corpo da função, use um ciclo `'for'` para percorrer a lista de gêneros errados, extrair a coluna `'genre'` e aplicar o método `replace` para fazer as correções.

In [19]:
# função para substituir duplicados implícitos
def replace_wrong_genres(df, column, wrong_genres, correct_genre):
    for wrong_genre in wrong_genres:
        df[column] = df[column].replace(wrong_genre, correct_genre)
    return df


Agora, chame a função `replace_wrong_genres()` e passe argumentos apropriados para que ela limpe duplicados implícitos (`hip`, `hop` e `hip-hop`) substituindo-os por `hiphop`:

In [20]:
# removendo duplicados implícitos
wrong_genres = ['hip', 'hop', 'hip-hop']
correct_genre = 'hiphop'
df = replace_wrong_genres(df, 'genre', wrong_genres, correct_genre)

Certifique-se que os nomes duplicados foram removidos. Imprima a lista de valores únicos da coluna `'genre'` mais uma vez:

In [21]:
# verificando valores duplicados
print(df['genre'].unique())

<StringArray>
[       'rock',         'pop',        'folk',       'dance',      'rusrap',
      'ruspop',       'world',  'electronic',           nan, 'alternative',
 ...
     'indipop',         'axé',        'fado',   'showtunes',       'arena',
       'irish',    'mandopop',       'forró',       'dirty',    'regional']
Length: 266, dtype: str


[Voltar ao Índice](#back)

A análise de elementos duplicados deve ser feita de maneira rígida para evitar conclusões equívocas com os resultados das análises. Numa lista com índices tão grandes, pode ser um trabalho bem exaustivo, contudo necessário. A abordagem usada para eliminá-los é eficiente e simples de se aplicar, de modo que foi útil no processo de limpeza dos dados.

[Voltar ao Índice](#back)

## Etapa 3. Teste da hipótese <a id='hypothesis'></a>

### Hipótese: comparação do comportamento dos usuários nas duas cidades <a id='activity'></a>

A hipótese afirma que existem diferenças no consumo de música pelos usuários em Springfield e em Shelbyville. Para testar a hipótese, use os dados dos três dias da semana: segunda-feira (Monday), quarta-feira (Wednesday) e sexta-feira (Friday).

* Agrupe os usuários por cidade.
* Compare o número de músicas tocadas por cada grupo na segunda, quarta e sexta.


Execute cada cálculo separadamente.

O primeiro passo é avaliar a atividade dos usuários em cada cidade. Não se esqueça das etapas "divisão-aplicação-combinação" sobre as quais falamos anteriormente na lição. Agora seu objetivo é agrupar os dados por cidade, aplicar o método de contagem apropriado durante a etapa de aplicação e então encontrar o número de músicas tocadas por cada grupo, especificando a coluna para a qual você quer obter a contagem.

Veja um exemplo de como o resultado final deve ser:
`df.groupby(by='....')['column'].method()` Execute cada cálculo separadamente.

Para avaliar a atividade dos usuários em cada cidade, agrupe os dados por cidade e encontre o número de músicas reproduzidas em cada grupo.



In [22]:
# Contando as músicas tocadas em cada cidade
df.groupby(by='city')['genre'].count()

city
Shelbyville    18234
Springfield    41892
Name: genre, dtype: int64

Springfield possui mais que o dobro de ouvintes se comparado a Shelbyville.

Agora vamos agrupar os dados por dia da semana e encontrar a quantidade de músicas tocadas na segunda, quarta e sexta-feira. Use a mesma abordagem que antes, mas agora precisamos agrupar os dados de uma forma diferente.


In [23]:
# Calculando as músicas escutadas em cada um desses três dias
df.groupby(by='day')['city'].count()

day
Friday       21840
Monday       21354
Wednesday    18059
Name: city, dtype: int64

O dia com maior número de reproduções musicais foi sexta-feira, seguido por segunda-feira e quarta-feira

Você acabou de aprender como contar entradas agrupando-as por cidade ou por dia. E agora você precisa escrever uma função que possa contar entradas simultaneamente com base em ambos os critérios.

Crie a função `number_tracks()` para calcular o número de músicas tocadas em um determinado dia **e** em uma determinada cidade. A função deve aceitar dois parâmetros:

- `day`: um dia da semana pelo qual precisamos filtrar os dados. Por exemplo, `'Monday'`.
- `city`: uma cidade pela qual precisamos filtrar os dados. Por exemplo, `'Springfield'`.

Dentro da função, você vai aplicar uma filtragem consecutiva com indexação lógica.

Primeiro, filtre os dados por dia e então filtre a tabela resultante por cidade.

Depois de filtrar os dados usando os dois critérios, conte o número de valores na coluna 'user_id' da tabela resultante. O resultado da contagem representará o número de entradas que você quer encontrar. Armazene o resultado em uma nova variável e imprima-o.

In [24]:
# Declare a função number_tracks() com dois parâmetros: day= e city=.
def number_tracks(df, day, city):
    # Armazene as linhas do DataFrame em que o valor na coluna 'day' é igual ao parâmetro day=
    new_df = df[df['day'] == day]
    # Filtre as linhas em que o valor na coluna 'city' é igual ao parâmetro city=
    new_df = new_df[new_df['city'] == city]
    # Extraia a coluna 'user_id' da tabela filtrada e aplique o método count()
    column_id = new_df['user_id']
    # Retorne o número dos valores da coluna 'user_id'
    return column_id.count()

Chame a função `number_tracks()` seis vezes, mudando os valores dos parâmetros, para que você possa recuperar os dados de ambas as cidades para cada um dos três dias.

In [25]:
# a quantidade de músicas tocadas em Springfield na segunda-feira
total = number_tracks(df, 'Monday', 'Springfield')
print(total)

15740


In [26]:
# a quantidade de músicas tocadas em Shelbyville na segunda-feira
total = number_tracks(df, 'Monday', 'Shelbyville')
print(total)

5614


In [27]:
# a quantidade de músicas tocadas em Springfield na quarta-feira
total = number_tracks(df, 'Wednesday', 'Springfield')
print(total)

11056


In [28]:
# a quantidade de músicas tocadas em Shelbyville na quarta-feira
total = number_tracks(df, 'Wednesday', 'Shelbyville')
print(total)

7003


In [29]:
# a quantidade de músicas tocadas em Springfield na sexta-feira
total = number_tracks(df, 'Friday', 'Springfield')
print(total)

15945


In [30]:
# a quantidade de músicas tocadas em Shelbyville na sexta-feira
total = number_tracks(df, 'Friday', 'Shelbyville')
print(total)

5895


**Conclusões**

A hipótese de que a atividade dos usuários é diferente a depender do dia da semana e cidade são verdadeiras.

Após síntese de todos os dados recolhidos e processados, é factível a conclusão que especialmente ás quartas-feira, a atividade é ligeiramente superior a média na cidade de Shelbyville, enquanto na cidade de Springfield ocorre o fenômeno inverso.

Nos demais dias, é possível reparar em que a oscilação de atividade é quase imperceptível, com a cidade de Springfield com índices ligeiramente superiores em relação a sua média de atividade, enquanto em Shelbyville os dados são ligeiramente inferiores.

reproduções:    Springfield  42741
                media        14247
                Shelbyville  18512
                media        6170
                razão        2,3
              
segunda-feira:  Springfield  15740
                Shelbyville  5614
                razão        2,77

quarta-feira:   Springfield  11056
                Shelbyville  7003
                razão        1,57

sexta-feira:    Springfield   15945
                Shelbyville   5895
                razão         2,7

[Voltar ao Índice](#back)

# Conclusões <a id='end'></a>

A hipótese apresentada foi muito pertinente. Foi possível observar através da análise das informações colhidas as diferenças de atividade no streaming a depender do dia da semana e da cidade.
Com estas informações, o streaming pode desenvolver estratégias específicas e planejadas para incentivar aumento de consumo nos dias de menor atividade.

[Voltar ao Índice](#back)